Interactivity API: Convert types of generator functions to async functions

luisherranz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This pull request updates the store return types to have their generators converted into async functions.

This is still a work in progress, as I would like to find a way to properly test these types before merging this change.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Even though developers have to use generators to create asynchronous functions in the stores, the store proxy returns an async function, not a generator. So when those actions are used as async functions, TypeScript still considers them as generators.

The async function is created on the fly by the store proxy: 

https://github.com/WordPress/gutenberg/blob/0d1943cf84e60ad4646333ed68eca7cc7b879009/packages/interactivity/src/store.ts#L108-L109

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

With a utility that checks if the type corresponds to a generator and if so, changes it to an async function recursively.

luisherranz committed:
Convert types of generator functions to async functions in stores

sirreal commented on packages/interactivity/src/store.ts:
```diff
@@ -207,6 +207,24 @@ interface StoreOptions {
+// Utility type to check if a function is a generator function.
+type IsGenerator< T > = T extends (
+	...args: infer A
+) => Generator< infer R, any, any >
+	? ( ...args: A ) => Promise< R >
+	: never;
```
This doesn't look quite right, the generators should _yield_ promises, but I think the async function this transforms to should resolve to the returned value. Probably something more like this:

```suggestion
) => Generator< any, infer R, any >
	? ( ...args: A ) => Promise< R >
	: never;
```

The first type argument could be `Promise< any >`, but that's probablye too strict. We await the yielded values and you can await values of any type (although it only makes sense for promises).

The name `IsGenerator` is a bit strange because this actually does the conversion.

luisherranz replied:
Indeed, thank you 🙂

Fixed in https://github.com/WordPress/gutenberg/pull/62400/commits/5c73a529452d0b8c82ec0dc1827681e5a198ead2.

luisherranz committed:
Fix return type

luisherranz committed:
Add types tests

luisherranz committed:
Refactor tests and update tsconfig.json to disable noUnusedLocals rule

luisherranz committed:
Move the tests to the `test-types` folder

luisherranz committed:
Add types test runner

luisherranz committed:
Add GH action

luisherranz committed:
Setup Node.js and install dependencies

luisherranz commented:
In the end, I ended up adding a system to test types in Gutenberg, including a npm script and a GitHub Action.

This is how it works:

<div>
    <a href="https://www.loom.com/share/8394c04f3273422cbaf967db0fb42923">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/8394c04f3273422cbaf967db0fb42923-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/8394c04f3273422cbaf967db0fb42923

---

The tests I have added to verify that the types of the Store and the Interactivity API work correctly are still very basic, but I will add more tests to cover all possibilities before merging this pull request.

luisherranz committed:
Fix test and add note

sirreal commented:
Rather than rolling our own type testing, [let's evaluate the options listed here](https://github.com/microsoft/DefinitelyTyped-tools/tree/main/packages/dtslint#just-looking-for-a-type-testing-tool):

> If you are just looking for a TypeScript type testing tool, use:
> 
> - [tsd](https://github.com/tsdjs/tsd)
> - [eslint-plugin-expect-type](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type)
> - [tstyche](https://github.com/tstyche/tstyche)

luisherranz commented:
Oh, nice. Can any of those be integrated with the existing Jest setup?

sirreal commented:
I'm not directly familiar with any of them, we'd need to evaluate which is best.

It looks like at least one integrates with eslint and uses special comments, so that might be something to add directly to source files and errors would appear in linting. Another seems to extend jest to check types so that would be part of unit tests.

mirka commented:
> Rather than rolling our own type testing, [let's evaluate the options listed here](https://github.com/microsoft/DefinitelyTyped-tools/tree/main/packages/dtslint?rgh-link-date=2024-07-02T09%3A55%3A51Z#just-looking-for-a-type-testing-tool):

For what it's worth, in the components package we've simply been [appending](https://github.com/search?q=repo%3AWordPress%2Fgutenberg+%22static+typ%22+path%3A%22%2Ftest%2F%22&type=code) a "static type checks" section to our unit tests, which has been good enough for our needs. We only feel the need for type tests when there are complex conditionals involved, so not often.

luisherranz commented:
We'll try that. Thanks @MIRKA!

luisherranz committed:
Merge branch 'trunk' into interactivity-api-types-transform-generators-into-async-functions

luisherranz committed:
Add the TS tests to the unit tests file

luisherranz commented:
@mirka, I have added the type tests, including a failing test, to a unit test file, but I don't see the error being reported in any way. Neither by Jest when you run the tests, nor when you do `npm run dev` in the build process or later in the watch. 

This is the file: https://github.com/WordPress/gutenberg/blob/bd2eaa4a08eb7ddfec955422442c60b4928efd15/packages/interactivity/src/test/store.ts

What is your workflow to see that a test is failing? Or do I have to add that unit test file somewhere?

luisherranz committed:
Add eslint disablers

luisherranz committed:
Add eslint disablers for old file

mirka commented:
> @mirka, I have added the type tests, including a failing test, to a unit test file, but I don't see the error being reported in any way. Neither by Jest when you run the tests, nor when you do `npm run dev` in the build process or later in the watch.

It looks like you're set up correctly. The type tests are static, so problems will be caught by Eslint. On CI, that runs in "Static Analysis". So if your passing test case starts to fail, or if the `expect-error` test case stops erroring, Eslint can flag it both locally (e.g. in IDE, `lint-staged`) and on CI. Is that not what you're seeing?

luisherranz commented:
That's not what I'm seeing, indeed. I have included a test that should fail [here](https://github.com/WordPress/gutenberg/blob/433cd28795c6d1e528105cf00c7946af0edbb7e3/packages/interactivity/src/test/store.ts#L35-L37), but as you can see in the GitHub Actions, the Static Analysis has passed correctly. It also didn't give me any warnings when committing (`lint-staged`).

luisherranz commented:
I think we also need to convert the input types so that people can define their actions using asynchronous functions, and these, internally, are compatible with generators.

Users should be able to define store types like this:

```ts
type MyStore = {
  actions: {
    myAsyncAction: () => Promise<void>;
  };
};

// This should be fine.
const { actions } = store<MyStore>('...', {
  actions: {
    *myAsyncAction() {
      // ...
    },
  },
});

// This should also be fine.
await actions.myAsyncAction();
```

luisherranz committed:
Prettify TS output to show the final types

luisherranz commented:
I just realized that I made some modifications from `github.dev` where the linter doesn't run, nor does the fork of Prettier, and I broke the Static Analysis test, but that has nothing to do with the previous issue, sorry. I'll fix it now from my local environment.

mirka commented:
> That's not what I'm seeing, indeed. I have included a test that should fail [here](https://github.com/WordPress/gutenberg/blob/433cd28795c6d1e528105cf00c7946af0edbb7e3/packages/interactivity/src/test/store.ts#L35-L37)

You're right, I can confirm that removing a `@ts-expect-error` from your tests don't fail the `tsc` build. Doing the same thing in a type test in the `@wordpress/components` _will_ break the build. So, some kind of tsconfig issue? (I'm testing with `npx tsc --build` by the way.)

luisherranz commented:
Replaced by #64577

I have moved the changes of TypeScript and the tests to that pull request, although they are not failing there either.

cc: @mirka
